kartothek/io_components/read.py

```diff
@@ -47,14 +48,21 @@ def dispatch_metapartitions_from_factory(
         if isinstance(ix, ExplicitSecondaryIndex)
     }
 
-    if concat_partitions_on_primary_index:
+    if dispatch_by or concat_partitions_on_primary_index:
         if dataset_factory.explicit_partitions:
             dataset_factory = dataset_factory.load_partition_indices()
 
+        if dispatch_by:
+            if not set(dispatch_by).issubset(dataset_factory.partition_keys):
+                raise ValueError("Nope.")
+            part_keys = dispatch_by
+        else:
+            part_keys = dataset_factory.partition_keys
+
         # Build up a DataFrame that contains per row a Partition and its
         # primary index columns.
         base_df = None
-        for part_key in dataset_factory.partition_keys:
+        for part_key in part_keys:
```

Comment:
Can you try to replace this loop with https://github.com/JDASoftwareGroup/kartothek/blob/a27ce44986985b9ed7e110472510d28c3cba7f70/kartothek/core/dataset.py#L339
Should go as
```
base_df = dataset_factory. get_indices_as_dataframe(columns=part_keys)
```
And see if this will do the trick (functionally)

Reply:
IF this is the case we need to look into performance since allowed_labels is not respected by get_indices_as_dataframe but let's first see if we can get rid of the duplication

Reply:
Will do that when the functional part is ready. Hope that your PR did most of the job... 😁